src/java.base/share/classes/java/lang/String.java

```diff
@@ -2500,8 +2501,9 @@ public int indexOf(int ch) {
      * {@code fromIndex} were larger than the string length, or were negative.
      */
     public int indexOf(int ch, int fromIndex) {
-        return isLatin1() ? StringLatin1.indexOf(value, ch, fromIndex, length())
-                : StringUTF16.indexOf(value, ch, fromIndex, length());
+        fromIndex = Math.max(0, fromIndex);
+        return isLatin1() ? StringLatin1.indexOf(value, ch, fromIndex, value.length)
+                : StringUTF16.indexOf(value, ch, fromIndex, value.length >> 1);
```

Comment:
This needs to include the check for `fromIndex >= this.length()`:
```suggestion
        fromIndex = Math.max(0, fromIndex);
        int toIndex = length();
        if (fromIndex >= toIndex) {
            return -1;
        }
        return isLatin1()
                ? StringLatin1.indexOf(value, ch, fromIndex, toIndex)
                : StringUTF16.indexOf(value, ch, fromIndex, toIndex);
```

Reply:
While not directly observable (all tests pass) then sadly the intrinsic implementation emits a string range check that triggers a decompilation and subsequent compilation without the intrinsic if you'd call with a `fromIndex` with values greater than `length()` (equal to should be fine). I'll try out some options here.